File: backend/app/routers/loja.py
"""Endpoints autenticados da loja — vendedor (anúncios/vendas) e comprador (pedidos)."""
from fastapi import APIRouter, Depends, HTTPException, Query

from app.dependencies import get_current_personal_id
from app.models.loja import (
    AvaliarBody,
    CapaUploadUrlBody,
    CriarAnuncioBody,
    CriarPedidoBody,
    EditarAnuncioBody,
)
from app.services import loja_service, media_service

router = APIRouter(prefix="/v1/loja", tags=["loja"])


# ── Vendedor: anúncios ────────────────────────────────────────────────────────

@router.post("/anuncios", status_code=201)
def criar_anuncio(body: CriarAnuncioBody, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.criar_anuncio(personal_id, body)


@router.patch("/anuncios/{anuncio_id}")
def editar_anuncio(anuncio_id: str, body: EditarAnuncioBody,
                   personal_id: str = Depends(get_current_personal_id)):
    return loja_service.editar_anuncio(personal_id, anuncio_id, body)


@router.get("/meus-anuncios")
def meus_anuncios(personal_id: str = Depends(get_current_personal_id)):
    return {"anuncios": loja_service.meus_anuncios(personal_id)}


@router.post("/anuncios/capa-upload-url")
def capa_upload_url(body: CapaUploadUrlBody,
                    personal_id: str = Depends(get_current_personal_id)):
    result = media_service.gerar_presigned_upload_url_loja(
        personal_id, body.filename, body.content_type
    )
    if not result:
        raise HTTPException(502, "Não foi possível gerar a URL de upload.")
    return result


# ── Comprador: pedidos ────────────────────────────────────────────────────────

@router.post("/pedidos", status_code=201)
def criar_pedido(body: CriarPedidoBody, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.criar_pedido(personal_id, body.anuncio_id)


@router.get("/pedidos/{pedido_id}")
def get_pedido(pedido_id: str, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.get_pedido(personal_id, pedido_id)


@router.post("/pedidos/{pedido_id}/instalar")
def instalar_pedido(pedido_id: str, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.instalar(personal_id, pedido_id)


@router.get("/pedidos/{pedido_id}/download")
def download_pedido(pedido_id: str, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.download(personal_id, pedido_id)


@router.post("/pedidos/{pedido_id}/cancelar")
def cancelar_pedido(pedido_id: str, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.cancelar_pedido(personal_id, pedido_id)


@router.get("/minhas-compras")
def minhas_compras(personal_id: str = Depends(get_current_personal_id),
                   limit: int = Query(default=50, ge=1, le=100),
                   cursor: str | None = None):
    return loja_service.minhas_compras(personal_id, limit, cursor)


# ── Vendedor: vendas ─────────────────────────────────────────────────────────

@router.get("/minhas-vendas")
def minhas_vendas(personal_id: str = Depends(get_current_personal_id),
                  limit: int = Query(default=50, ge=1, le=100),
                  cursor: str | None = None):
    return loja_service.minhas_vendas(personal_id, limit, cursor)


@router.post("/vendas/{pedido_id}/confirmar")
def confirmar_venda(pedido_id: str, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.confirmar_venda(personal_id, pedido_id)


@router.post("/vendas/{pedido_id}/cancelar")
def cancelar_venda(pedido_id: str, personal_id: str = Depends(get_current_personal_id)):
    return loja_service.cancelar_pedido(personal_id, pedido_id)


# ── Comprador: avaliações ────────────────────────────────────────────────────

@router.post("/anuncios/{anuncio_id}/avaliacoes")
def avaliar(anuncio_id: str, body: AvaliarBody,
            personal_id: str = Depends(get_current_personal_id)):
    return loja_service.avaliar(personal_id, anuncio_id, body)

File: frontend/src/api/loja.ts
import { api } from './client'

// ── Tipos (espelham backend/app/models/loja.py + loja_service) ───────────────

export type PedidoStatus =
  | 'AGUARDANDO_PAGAMENTO'
  | 'AGUARDANDO_VENDEDOR'
  | 'ENTREGUE'
  | 'EXPIRADO'
  | 'CANCELADO'

export type AnuncioStatus = 'PUBLICADO' | 'PAUSADO' | 'REMOVIDO_ADMIN'

export interface PacoteStats {
  n_exercicios: number
  n_templates: number
  n_rotinas: number
}

export interface AnuncioCard {
  anuncio_id: string
  titulo: string
  preco_centavos: number
  capa_url?: string | null
  capa_s3_key?: string | null
  vendedor_id: string
  vendedor_nome?: string
  stats: PacoteStats
  avaliacao_media: number | null
  avaliacao_count: number
  vendas_count: number
  criado_em?: string
}

export interface AnuncioDetalhe extends AnuncioCard {
  descricao: string
  status: AnuncioStatus
  pacote_id: string
  pacote_nome?: string
  pacote_versao?: string
  mp_disponivel?: boolean
  avaliacao_soma?: number
}

export interface Avaliacao {
  comprador_id: string
  comprador_nome?: string
  nota: number
  comentario?: string
  criado_em: string
  atualizado_em?: string
}

export interface PixData {
  payment_id: string
  qr_code?: string
  qr_code_base64?: string
  expires_at?: string
  status?: string
}

export interface Pedido {
  pedido_id: string
  anuncio_id: string
  titulo: string
  pacote_id: string
  preco_centavos: number
  vendedor_id: string
  vendedor_nome?: string
  comprador_id: string
  comprador_nome?: string
  modo: 'MP' | 'MANUAL' | 'GRATIS'
  status: PedidoStatus
  mp_payment_id?: string
  expira_em?: string
  entregue_em?: string
  criado_em: string
  pix?: PixData
}

export interface InstalarResponse {
  pacote_id: string
  nome?: string
  exercicios_importados?: number
  templates_importados?: number
  rotinas_importadas?: number
  licenciado?: boolean
}

export interface PacoteGerado {
  pacote_id: string
  nome: string
  descricao?: string
  versao?: string
  max_usos?: number
  n_exercicios?: number
  n_templates?: number
  n_rotinas?: number
  criado_em: string
}

export function formatPreco(centavos: number): string {
  return (centavos / 100).toLocaleString('pt-BR', { style: 'currency', currency: 'BRL' })
}

/** Rótulo de preço para exibição — pacotes com preço 0 são gratuitos (resgate). */
export function precoLabel(centavos: number): string {
  return centavos === 0 ? 'Grátis' : formatPreco(centavos)
}

// ── API ──────────────────────────────────────────────────────────────────────

export const lojaApi = {
  // Público (catálogo)
  catalogo: (cursor?: string | null) =>
    api
      .get<{ anuncios: AnuncioCard[]; cursor: string | null }>('/v1/public/loja/anuncios', {
        params: { limit: 24, cursor: cursor ?? undefined },
      })
      .then((r) => r.data),

  detalhe: (anuncioId: string) =>
    api.get<AnuncioDetalhe>(`/v1/public/loja/anuncios/${anuncioId}`).then((r) => r.data),

  avaliacoes: (anuncioId: string, cursor?: string | null) =>
    api
      .get<{ avaliacoes: Avaliacao[]; cursor: string | null }>(
        `/v1/public/loja/anuncios/${anuncioId}/avaliacoes`,
        { params: { cursor: cursor ?? undefined } },
      )
      .then((r) => r.data),

  // Comprador
  criarPedido: (anuncioId: string) =>
    api.post<Pedido>('/v1/loja/pedidos', { anuncio_id: anuncioId }).then((r) => r.data),

  getPedido: (pedidoId: string) =>
    api.get<Pedido>(`/v1/loja/pedidos/${pedidoId}`).then((r) => r.data),

  instalar: (pedidoId: string) =>
    api.post<InstalarResponse>(`/v1/loja/pedidos/${pedidoId}/instalar`).then((r) => r.data),

  download: (pedidoId: string) =>
    api
      .get<{ fmt: string; pacote_id: string; token: string; filename: string }>(
        `/v1/loja/pedidos/${pedidoId}/download`,
      )
      .then((r) => r.data),

  cancelarPedido: (pedidoId: string) =>
    api.post<Pedido>(`/v1/loja/pedidos/${pedidoId}/cancelar`).then((r) => r.data),

  minhasCompras: () =>
    api
      .get<{ compras: Pedido[]; cursor: string | null }>('/v1/loja/minhas-compras')
      .then((r) => r.data),

  avaliar: (anuncioId: string, nota: number, comentario: string) =>
    api
      .post(`/v1/loja/anuncios/${anuncioId}/avaliacoes`, { nota, comentario })
      .then((r) => r.data),

  // Vendedor
  meusAnuncios: () =>
    api.get<{ anuncios: AnuncioDetalhe[] }>('/v1/loja/meus-anuncios').then((r) => r.data),

  criarAnuncio: (body: {
    pacote_id: string
    titulo: string
    descricao: string
    preco_centavos: number
    capa_s3_key?: string | null
  }) => api.post<AnuncioDetalhe>('/v1/loja/anuncios', body).then((r) => r.data),

  editarAnuncio: (
    anuncioId: string,
    body: Partial<{
      titulo: string
      descricao: string
      preco_centavos: number
      capa_s3_key: string
      pacote_id: string
      status: 'PUBLICADO' | 'PAUSADO'
    }>,
  ) => api.patch<AnuncioDetalhe>(`/v1/loja/anuncios/${anuncioId}`, body).then((r) => r.data),

  capaUploadUrl: (filename: string, contentType: string) =>
    api
      .post<{ upload_url: string; s3_key: string }>('/v1/loja/anuncios/capa-upload-url', {
        filename,
        content_type: contentType,
      })
      .then((r) => r.data),

  minhasVendas: () =>
    api
      .get<{ vendas: Pedido[]; cursor: string | null }>('/v1/loja/minhas-vendas')
      .then((r) => r.data),

  confirmarVenda: (pedidoId: string) =>
    api.post<Pedido>(`/v1/loja/vendas/${pedidoId}/confirmar`).then((r) => r.data),

  cancelarVenda: (pedidoId: string) =>
    api.post<Pedido>(`/v1/loja/vendas/${pedidoId}/cancelar`).then((r) => r.data),

  pacotesGerados: () =>
    api.get<{ pacotes: PacoteGerado[] }>('/v1/pacotes/gerados').then((r) => r.data),
}
